cpp/src/utilities/type_dispatcher.hpp

```diff
@@ -115,7 +115,7 @@ namespace cudf {
 #pragma hd_warning_disable 
 #pragma nv_exec_check_disable
 template <class functor_t, typename... Ts>
-CUDA_HOST_DEVICE_CALLABLE decltype(auto) type_dispatcher(gdf_dtype dtype,
+CUDA_HOST_DEVICE_CALLABLE constexpr auto type_dispatcher(gdf_dtype dtype,
```

Comment:
The `decltype(auto)` is necessary to allow returning references https://isocpp.org/wiki/faq/cpp14-language#decltype-auto

Reply:
Done.